Frontend/student-hub/src/components/ResetPassword.js

```diff
@@ -0,0 +1,181 @@
+            setError(error);
+          }
+        } catch (error) {
+          setError('Request failed. Please try again.')
+        }
+        setIsLoading(false);
+      };
+
+      // reset password
+      const handleResetPassword = async (e) => {
+        e.preventDefault();
+        setError('');
+        setMessage('');
+        setIsLoading(true);
+
+        try {
+          const response = await fetch(`${process.env.REACT_APP_API}/auth/reset-password`, {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ email: passwordResetEmail, code, newPassword, confirmPassword }),
+          });
+
+
+          if (response.ok) {
+            const data = await response.json();
+            setMessage(data.message);
+            navigate('/login')
+          } else {
+            const { error } = await response.json();
+            setError(error);
+          }
+        } catch (error) {
+          setError('Reset failed. please try again.');
+        }
+        setIsLoading(false);
+      };
+
+      // request password reset
+      const handleRequestResetPassword = async (e) => {
+        e.preventDefault();
+        setError('');
+        setMessage('');
+        setIsLoading(true);
+        try {
+          const response = await fetch(`${process.env.REACT_APP_API}/auth/password-resetcode`, {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json'},
+            body: JSON.stringify({ email: passwordResetEmail }),
+          });
+
+          if (response.ok) {
+            const data = await response.json();
+            setMessage(data.message);
+          } else {
+            const { error } = await response.json();
+            setError(error);
+          }
+        } catch (error) {
+          setError('Request failed. Please try again.');
+        }
+        setIsLoading(false);
+      };
+
+      return (
+        <Container maxWidth="sm">
+          <Grid container spacing={2} alignItems="center" justifyContent="center" style={{ minHeight: '80vh' }}>
+            <Grid item xs={12}>
+              <h2>{message ? "Reset Password" : "Request Password Reset"}</h2>
+            </Grid>
+            <Grid item xs={12}>
+              <form onSubmit={message ? handleResetPassword : handleRequestResetPassword}>
+                <Grid container spacing={2}>
+                  {!message && (
+                    <Grid item xs={12}>
+                      <TextField
+                        fullWidth
+                        type="email"
+                        label="Email"
+                        variant="outlined"
+                        value={passwordResetEmail}
+                        onChange={(e) => setPasswordResetEmail(e.target.value)}
+                        required
+                      />
+                    </Grid>
+                  )}
+                  {message && (
```

Comment:
it feels a bit off to me that we're checking whether message is empty or not in order to switch between the two forms.

I think rather than checking if message is empty, it's better to instead have a button to switch between the two forms which you can map to a new state - something like "Already requested a code? Click here to enter the code". That way, users can freely switch between the two forms without relying on only being able to enter the code if they've just requested a code.

Reply:
Before switching between the two forms, the user must enter their email address to receive a new code. This code is sent to their email and expires every 1 minute. If the user requests a code but doesn't enter it, it will expire, and they'll need to request a new one. The email verification step ensures that the user is the account owner, preventing someone from gaining access to someone else's account, even if they obtain the code by chance.